_posts/2021-02-16-message-passing.md

```diff
@@ -0,0 +1,209 @@
+---
+
+Message passing has always been central to Erlang, and while reasonably
+well-documented we've avoided going into too much detail to give us more
+freedom when implementing it. There's nothing preventing us from describing it
+in a blog post though, so let's have a closer look!
+
+Erlang processes communicate with each other by sending each other _signals_
+(not to be confused with Unix signals). There are many different kinds and
+_messages_ are just the most common one. Practically everything involving more
+than one process uses signals internally: for example, the `link/1` function is
+implemented by having the involved processes talk back and forth until they've
+agreed on a link.
+
+This helps us avoid a great deal of locks and would make an interesting blog
+post on its own, but for now we only need to keep two things in mind: all
+signals (including messages) are continuously received and handled behind the
+scenes, and they have a [defined order]:
+
+Signals between two processes are guaranteed to arrive in the order they were
+sent. In other words, if process `A` sends signals `1` and `2` to process `B`,
+signal `1` is guaranteed to arrive before signal `2`.
+
+Why is this important? Consider the request-response idiom:
+
+```erlang
+%% Request a 'DOWN' message when the process identified by `Pid` dies.
+Mref = monitor(process, Pid),
+%% Send a message to `Pid`.
+Pid ! {self(), Mref, Request},
+receive
+    {Mref, Response} ->
+        %% Undo monitor/2 and remove any 'DOWN' messages that might have
+        %% arrived in the meantime.
+        erlang:demonitor(Mref, [flush]),
+        {ok, Response};
+    {'DOWN', Mref, _, _, Reason} ->
+        {error, Reason}
+end
+```
+
+Since dead processes cannot send messages we know that the response must come
+before any eventual `'DOWN'` message, but without a guaranteed order the
+`'DOWN'` message could arrive before the response and we'd have no idea whether
+a response was coming or not, which would be very annoying to deal with.
+
+This saves us quite a bit of hassle and doesn't come at much of a cost, but
+the guarantees stop there. If more than one process sends messages to a common
+process, they can arrive in any order even when you "know" that one of the
+messages was sent first. For example, this sequence of events is legal and
+entirely possible:
+
+1. `A` sends message `1` to `B`
+1. `A` sends message `2` to `C`
+1. `C`, in response to message `2`, sends message `3` to `B`
+1. `B` receives message `3`
+1. `B` receives message `1`
+
+Luckily, global orders are rarely needed and are easy to impose yourself
+(outside distributed cases): just let all involved parties synchronize with a
+common process.
+
+### Sending messages
+
+Sending a message is straightforward: we try to find the process associated
+with the process identifier, and if one exists we insert the message into its
+signal queue.
+
+Messages are always copied before being inserted into the queue. As wasteful as
```

Comment:
The copying is conceptual and in the current implementation it is actually a real copying going on. Optimizations and GC strategies might try to avoid real copying if it is not strictly necessary.

Reply:
I've reworded it a bit now, comparing it to non-copying implementations and explained why we didn't go down that path.